data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestHandler.java

```diff
@@ -149,61 +144,46 @@ public void handle(final HttpServerRequest request) {
   }
 
   @Override
-  public Future<Void> reconcile(Map<Resource, Set<Egress>> resourcesMap) {
-    final Map<String, Entry<String, Producer<K, V>>> newProducers = new HashMap<>();
-    final var producers = this.producers.get();
-
-    for (final var resource : resourcesMap.keySet()) {
-      //TODO support host ingress too
-      //TODO check if there is an ingress, otherwise fail
-      final var pair = producers.get(resource.ingress().getPath());
-
-      if (pair == null) {
-        // There is no producer for this Resource, so create it and add it to newProducers.
-        addResource(newProducers, resource);
-        continue;
-      }
-
-      if (!pair.getKey().equals(resource.bootstrapServers())) {
-        // Bootstrap servers changed, close the old producer, and re-create a new one.
-        final var producer = pair.getValue().producer;
-        producer.flush(complete -> producer.close());
-
-        addResource(newProducers, resource);
-        continue;
-      }
-
-      // Nothing changed, so add the previous entry, to newProducers.
-      newProducers.put(resource.ingress().getPath(), pair);
-    }
-
-    this.producers.set(newProducers);
-
-    logger.debug("Added resources to handler {}", keyValue("resources", newProducers.keySet()));
-
-    return Future.succeededFuture();
-  }
-
-  private void addResource(
-    final Map<String, Entry<String, Producer<K, V>>> producers,
-    final Resource resource) {
-
+  public Future<Void> onNewIngress(
+    DataPlaneContract.Resource resource,
+    DataPlaneContract.Ingress ingress) {
     final var producerConfigs = (Properties) this.producerConfigs.clone();
-    producerConfigs.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, resource.bootstrapServers());
-    if (resource.ingress().getContentMode() != DataPlaneContract.ContentMode.UNRECOGNIZED) {
-      producerConfigs.setProperty(CloudEventSerializer.ENCODING_CONFIG, encoding(resource.ingress().getContentMode()));
+    producerConfigs.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, resource.getBootstrapServers());
+    if (ingress.getContentMode() != DataPlaneContract.ContentMode.UNRECOGNIZED) {
+      producerConfigs.setProperty(CloudEventSerializer.ENCODING_CONFIG, encoding(ingress.getContentMode()));
     }
     producerConfigs.setProperty(CloudEventSerializer.EVENT_FORMAT_CONFIG, JsonFormat.CONTENT_TYPE);
 
     final KafkaProducer<K, V> producer = producerCreator.apply(producerConfigs);
 
-    producers.put(
-      resource.ingress().getPath(),
+    this.producers.get().put(
+      ingress.getPath(),
       new SimpleImmutableEntry<>(
-        resource.bootstrapServers(),
-        new Producer<>(producer, resource.topics().iterator().next())
+        resource.getBootstrapServers(),
+        new Producer<>(producer, resource.getTopics(0))
       )
```

Comment:
The map inside the `AtomicReference` isn't thread-safe.
We shouldn't change the map to be a synchronized one since it might impact performance in the hot path and so
we need to create a new map based on the current map we hold and swap it at the end of the reconcile loop.

Reply:
This is in fact related to https://github.com/knative-sandbox/eventing-kafka-broker/issues/332. Because to recreate the same pattern we used before i need to perform several changes here (most prominently, add the concept of state to the reconcilers and the listeners), I prefer to use a synchronized map temporarly now and then, with #332 in, we can use a normal map here too

Reply:
Even better, i can keep this single lock using cas, then we reiterate here with #332

Reply:
Check it out now